### ISSUE TITLE
is there a usecase for nodeify ?

### ISSUE THREAD

**Issue by qballer** [garronej/denoify#6]
Considering writing a package which does the other way around, thing is, I don't have a use case for that. Would it make sense to take a deno module and do what ever is needed to run in node ?

**Comment by garronej** [comment]
Hi @qballer,
There are definitively tons of use cases for Nodeify!  
It would be great when writing a module for Deno to also be able to support Node.   
It would be a step toward keeping the JavaScript community united. Peoples that develop for Node can bring Deno compact thanks to Denoify and peoples that develop for Deno could bring Node support with Nodeify.  

That said there are some things you might want to be aware before investing time in such a project:
1) Nodeify will be much harder to build than Denoify. Denoify leverage the [the Node compatibility layer](https://deno.land/std/node) which is a standard Deno module. For now, it's implementation is very incomplete but the community behind Deno will eventually complete the polyfills. There is no Deno compatibility layer for Node, you would need to write it...
2) Fixing the import statements is the easy part, handling dependencies is much more tricky. 
3) Once you have generated the source you need to provide a straightforward guide to explain how to publish on both deno.land/x and NPM. For most user tracking `dist` files on the main branch will not be an option.  

Long story short: I am pretty sure people would use it if you manage to pull it off but pulling it off won't be easy.

**Comment by qballer** [comment]
1. Makes sense. Can you elaborate more on what a node-comparability-layer will need in your mind? 
2. Which issues do you see with dev-tools? 
3. Shouldn't be a problem because I'm dealing with dev-tools for a while.

**Comment by garronej** [comment]
1. Well, you would need to polyfill for Node all of [thoes](https://deno.land/std). Meaning that you would have to create a node module for each entry in the list. Every module would have exactly the same API that the one defined in deno.land/std. Fun task but lots of work ahead.  
2. To handle dependencies you have two options: the first one is to bundle the URL import you ran against with "nodified" version of the files. To work this would require that Nodefiy work out of the box with any module, this is not realistic IMO.  
The other way to go would be to do like denoify, to require people to Nodeify their dependencies manually. Your script would need to generate a package.json... Not impossible but it needs to be carefully thought out.

**Comment by qballer** [comment]
thank a lot for the insights, I've decided to go for the project. If you would like to collab on it in anyway do let me know.

**Comment by garronej** [comment]
I can't invest dev time into Nodeify, I have already a lot on my plate with Denoify EVT and my work but if you want to consult me on design decision I will be happy to help.  
Once you get to the point where you have decent documentation and a couple of practical examples of Nodeify working on real modules I will be happy to point people to Nodeify in the Denoify documentation.  

It would also be great if you could align with Denoify with as many design decisions as possible. This could make our project the two faces of the same coin.  

One last piece of advice, keep your priority in check, focus on putting together a proof of concept. Maybe target a popular Deno module like for example [coward](https://github.com/fox-cat/coward) and see what needs to be done to support it.  

Good luck and have fun. 🚀

**Comment by garronej** [comment]
@qballer When picking [coward](https://github.com/fox-cat/coward) at random I didn't realize that it uses [EVT](https://evt.land), a package that uses Denoify.  
This makes me think that you should be able to automatically support 'Denoifyed' module by using their NPM counterpart. 

For example when you run against an import like:  
```typescript
export { Evt } from "https://deno.land/x/evt@1.7.12/mod.ts";
```
you should replace the statement by:

```typescript
export { Evt } from "evt";
```
and populate the generated ``package.json`` with:

```json
{
  "dependencies": {
    "evt": "1.7.12"
  }
}
```

To test if a module is "denoified" just look in the [`mod.ts`](https://raw.githubusercontent.com/garronej/evt/1.7.13/mod.ts) file for the `denoify` keyword and get the NPM module name in the `[package.json](https://raw.githubusercontent.com/garronej/evt/1.7.13/package.json)`.  

I, on the other end, should support by default Nodified package as well.  

Let us help each other out and build something great 🚀

**Comment by garronej** [comment]
Also be aware that you can run against specific imports like: 

```typescript
import { chunksOf } from "https://raw.githubusercontent.com/garronej/evt/1.7.13/lib/util/genericOperators/chunksOf.ts";
```

That should be replaced by

```typescript
import { chunksOf } from "evt/lib/util/genericOperators/chunksOf";
```

**Comment by garronej** [comment]
...And for modules deployer without using denoify_ci you could run against import like:

```typescript
import { Cat } from "https://raw.githubusercontent.com/garronej/my_dummy_npm_and_deno_module/0.2.5/deno_dist/lib/Cat.ts"
```

Thant should be replaced by (`deno_dist`->`dist`): 

```typescript
import { Cat } from "my_dummy_npm_and_deno_module/dist/lib/Cat.ts";
```

```json
{
  "dependencies": {
    "my-dummy-npm-and-deno-module": "0.2.5"
  }
}
```

**Comment by qballer** [comment]
I totally agree we should play well together. I thought about starting first with oak, a denofied package is an interesting edge case here. If you would like we can keep this issue open for further coordination.

**Comment by qballer** [comment]
Just to let you know I've dropped this project idea since it's too much work and I see little user base for it.

**Comment by wojpawlik** [comment]
Today I wrote https://github.com/wojpawlik/deno2node 🎉

**Comment by GervinFung** [comment]
Also there is [this](https://github.com/denoland/dnt)
